### ISSUE TITLE
Parallelize build

### ISSUE THREAD

**Issue by ayberkt** [agda/agda#6861]
I generated this [Makefile for TypeTopology](https://github.com/martinescardo/TypeTopology/blob/master/Makefile) by translating the dependency graph generated by Agda. This allows the build to be parallelized using the [parallel execution](https://www.gnu.org/software/make/manual/html_node/Parallel.html) feature of `make`.

However, this results in `agda` being called for each module, which is a pretty inefficient approach since it results in `.agdai` files being read over and over again.

I think it would be much more efficient to implement this as a feature in Agda.

**Comment by felixwellen** [comment]
I did the same thing, but generated the dependencies myself - didn't know Agda can do this. When I started, @andreasabel started to write something in Agda, but I don't know what the status of that is.
Here is my take on that:
https://github.com/agda/cubical/pull/1057
(which makes cubical type check in around 2/3 of the usual time, if you happen to have at least 32GB of ram...)

**Comment by ayberkt** [comment]
Thanks for the the information @felixwellen.

I wonder how an implementation of this in Agda itself compare in terms of efficiency. One could at least try the simple-minded approach of implementing exactly what `make` does in Agda itself. This should be fairly straightforward in my opinion, though I’m not familiar with the internals of Agda so I might be underestimating some technical challenges.

@andreasabel: What was the approach you had in mind for this?

**Comment by jwaldmann** [comment]
> ... which is a pretty inefficient approach since it results in `.agdai` files being read over and over again.

just looking at this as a by-stander: have you measured this? is parsing the bottleneck? or accessing the disk?  (after reading a file for the first time, the OS (the disk driver) should have the file contents cached?)

ghc/cabal currently has a similar discussion (well, with the same goal of parallelizing builds) https://well-typed.com/blog/2023/08/reducing-haskell-parallel-build-times/

**Comment by martinescardo** [comment]
I would be interested in these ideas being pursued further.

**Comment by felixwellen** [comment]
> Thanks for the the information [@felixwellen](https://github.com/felixwellen).
> 
> I wonder how an implementation of this in Agda itself compare in terms of efficiency. One could at least try the simple-minded approach of implementing exactly what `make` does in Agda itself. This should be fairly straightforward in my opinion, though I’m not familiar with the internals of Agda so I might be underestimating some technical challenges.
> 
> [@andreasabel](https://github.com/andreasabel): What was the approach you had in mind for this?

It is actually not straight forward to do this in a performant way. I used things like sed and grep to get all the dependencies and make to work with the dependency graph. At least some, and maybe all of these things would be a lot slower in agda if you work in a straight forward way (e.g. using agda's parser to figure out dependecies).

**Comment by felixwellen** [comment]
> I would be interested in these ideas being pursued further.

Good to know!
How much improvement in type-checking performance would make a real difference for you?

**Comment by martinescardo** [comment]
You said in another issue that you reduced the time from 30min to 17min in the cubical library. Such an improvement would also be good for TypeTopology.

**Comment by marcinjangrzybowski** [comment]
BTW: If you need some statistics to be gatehred I recently build a setup based on open-search to gather and analyse stats during compilation of larg codebases, let me know if you can think about some stat wich could be informative

**Comment by martinescardo** [comment]
I had a student project that implemented the creation of a makefile and then use this to get parallelism. There was some gain in performance. But the fact that the `agdai` files are read and parsed again and again makes this not as good as one could hope, if Agda implemented the parallelism directed and kept the `agdai` information in ram during the process.

**Comment by martinescardo** [comment]
Here is his project: https://github.com/JulianVidal/AGDA_HTML

**Comment by martinescardo** [comment]
I will ask him to include the pdf of his dissertation.

**Comment by martinescardo** [comment]
For example, in a Mac Mini M4 with 24GB of ram, TypeTopology went down from 345s to 265s (77% of the original time) with one of the various parallelizing strategies he tried. 

He also tried what he called the "unsafe" strategy, namely start the typechecking of all files in parallel, leaving the operating system to schedule the parallelism. I don't know how safe or usafe this is, but, it gives a reduction to 50% of the original time. in this machine.

**Comment by martinescardo** [comment]
> BTW: If you need some statistics to be gatehred I recently build a setup based on open-search to gather and analyse stats during compilation of larg codebases, let me know if you can think about some stat wich could be informative

Do you have a link to this?

**Comment by martinescardo** [comment]
Actually, Julian Vidal's dissertation is already there: https://github.com/JulianVidal/AGDA_HTML/blob/main/reports/Final_Report/Dissertation.pdf

**Comment by marcinjangrzybowski** [comment]
> Do you have a link to this?

it involves custom modifications to agda, and aws-hosted system for analytics, I could push it to custom branch, but I think that it might be easier if I gather data for you if I know what is of interrest, result is querrable and interactive, 
for now I was using it to get some insight about parts of cubical library suffering from lossy-unification requirements, I could reconfigure/modify it to gather stats relevant for this issue

**Comment by marcinjangrzybowski** [comment]
for starters I could check how much overhead comes from coldstarting agda every time, also there are some metrics mentioned in this conversations wich I can try to obtain

**Comment by martinescardo** [comment]
Yes, this would be useful!

**Comment by ayberkt** [comment]
Fixed in PR #8332.
